contrib/platform/src/com/sun/jna/platform/linux/LibC.java

```diff
@@ -0,0 +1,76 @@
+/* Copyright (c) 2017 Daniel Widdis, All Rights Reserved
+ * 
+ * The contents of this file is dual-licensed under 2 
+ * alternative Open Source/Free licenses: LGPL 2.1 or later and 
+ * Apache License 2.0. (starting with JNA version 4.0.0).
+ * 
+ * You can freely decide which license you want to apply to 
+ * the project.
+ * 
+ * You may obtain a copy of the LGPL License at:
+ * 
+ * http://www.gnu.org/licenses/licenses.html
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "LGPL2.1".
+ * 
+ * You may obtain a copy of the Apache License at:
+ * 
+ * http://www.apache.org/licenses/
+ * 
+ * A copy is also included in the downloadable source code package
+ * containing JNA, in file "AL2.0".
+ */
+package com.sun.jna.platform.linux;
+
+import com.sun.jna.Library;
+import com.sun.jna.Native;
+import com.sun.jna.NativeLong;
+import com.sun.jna.Structure;
+import com.sun.jna.Structure.FieldOrder;
+import com.sun.jna.platform.unix.LibCAPI;
+
+/**
+ * <I>libc</I> API
+ * 
+ * @author Daniel Widdis
+ */
+public interface LibC extends LibCAPI, Library {
+    String NAME = "c";
+    LibC INSTANCE = Native.load(NAME, LibC.class);
+
+    @FieldOrder({ "uptime", "loads", "totalram", "freeram", "sharedram", "bufferram", "totalswap", "freeswap", "procs",
+            "totalhigh", "freehigh", "mem_unit", "_f" })
+    class Sysinfo extends Structure {
+        public NativeLong uptime; // Seconds since boot
+        // 1, 5, and 15 minute load averages
+        public NativeLong[] loads = new NativeLong[3];
+        public NativeLong totalram; // Total usable main memory size
+        public NativeLong freeram; // Available memory size
+        public NativeLong sharedram; // Amount of shared memory
+        public NativeLong bufferram; // Memory used by buffers
+        public NativeLong totalswap; // Total swap space size
+        public NativeLong freeswap; // swap space still available
+        public short procs; // Number of current processes
+        public NativeLong totalhigh; // Total high memory size
+        public NativeLong freehigh; // Available high memory size
+        public int mem_unit; // Memory unit size in bytes
+        // Padding is based on the size of NativeLong. When the size is 4 we
+        // need 8 bytes of padding. When this size is 8, zero padding is needed
+        // but we are not allowed to declare an array of size zero. It's safe to
+        // declare extra padding in the structure; these bytes simply will not
+        // be written on 64-bit systems.
+        public byte[] _f = new byte[8]; // padding to 64 bytes
```

Comment:
You _could_ dynamically set this based on Native.LONG_SIZE, although I'm not entirely certain zero-length arrays handling is bug-free.

Reply:
Would it work to put a conditional into the FieldOrder and omit the `_f` if the padding is 0?  It presently throws an exception if the padding calculation results in "`new byte[0]`"

Reply:
zero-length arrays are not supported by JNA. There is a guard in `Native#getNativeSize` that rejects zero length arrays. From what I read, C does not allow for zero length arrays, so the JNA bindings is in line.

But in this case the array is just padding - what I would suggest:

- allow `com.sun.jna.Structure.calculateSize(boolean force, boolean avoidFFIType)` to be overriden by subclasses
- override it in `Sysinfo` like so:
```java
        @Override
        protected int calculateSize(boolean force, boolean b) {
            int superValue = super.calculateSize(force, b);
            if(superValue == CALCULATE_SIZE) {
                return superValue;
            } else {
                return superValue + (20 - 2 * NativeLong.SIZE - 4);
            }
        }
```

@twall what do you think? I aggree, that for a single structure you could just over allocate, but if for a reason the structure is in an array, this might break.

Reply:
Could you also do:
```
@FieldOrder(NativeLong.SIZE < 8 ? 
    { "uptime", "loads", "totalram", "freeram", "sharedram", "bufferram", "totalswap", 
             "freeswap", "procs", "totalhigh", "freehigh", "mem_unit", "_f" } :
    { "uptime", "loads", "totalram", "freeram", "sharedram", "bufferram", "totalswap",
             "freeswap", "procs", "totalhigh", "freehigh", "mem_unit" })
``` 
?

Reply:
Is the field _ever_ used?  If not, I'd say go with @matthiasblaesing 's suggestion. 

I don't think the field order tracking and public fields were ever designed to potentially have different lists, so dynamically setting the field order would require extra testing.

Reply:
I don't think you need to expose `calculateSize(boolean, boolean)`; `[calculateSize(boolean)](https://github.com/java-native-access/jna/blob/master/src/com/sun/jna/Structure.java#L1102)` is already exposed for overriding.

Reply:
The only concern about allocation size is if this structure was in an array.  `Structure.toArray()` calls `ensureAllocated()` which then calls (private) `ensureAllocated(false)` which eventually calls (package private) `calculateSize(true, avoidFFIType)`.  So I can override `calculateSize(boolean)` but it is never called in `toArray()`.

Reply:
I see what you're saying.  I can't think of a different refactor that would solve it, so `calculateSize(boolean,boolean)` should be `protected` instead of package private.